bundles/admin/admin-hierarchical-layerlist/components/Layer.js

```diff
@@ -0,0 +1,246 @@
+            });
+        });
+    },
+
+    /**
+     * Get mapl layer groups for layer group selection
+     * @method  _getMaplayerGroups
+     * @return  {Array}           groups
+     * @private
+     */
+    _getMaplayerGroups: function() {
+        var me = this;
+        var groups = [];
+        me.layerService.getAllLayerGroups().forEach(function(group) {
+            groups.push({
+                id: group.id,
+                cls: 'group',
+                name: Oskari.getLocalized(group.name)
+            });
+
+            // subgroups
+            group.groups.forEach(function(subgroup) {
+                groups.push({
+                    id: subgroup.id,
+                    cls: 'subgroup',
+                    name: Oskari.getLocalized(subgroup.name)
+                });
+
+                // subgroup subgroups
+                subgroup.groups.forEach(function(subgroupsubgroup) {
+                    groups.push({
+                        id: subgroupsubgroup.id,
+                        cls: 'subgroupsubgroup',
+                        name: Oskari.getLocalized(subgroupsubgroup.name)
+                    });
+                });
+            });
+        });
+
+        return groups;
+    },
+
+    /*******************************************************************************************************************************
+    /* PUBLIC METHODS
+    *******************************************************************************************************************************/
+    /**
+     * Get dataproviders
+     * @method getDataproviders
+     * @param  {Boolean}        emptyCache empty cache
+     * @param  {Function}       callback   callback function
+     */
+    getDataproviders: function(emptyCache, callback) {
+        var me = this;
+        if (me.dataProviders === null || emptyCache === true) {
+            jQuery.ajax({
+                type: 'GET',
+                dataType: 'json',
+                contentType: 'application/json; charset=UTF-8',
+                url: me.sandbox.getAjaxUrl('GetMapLayerGroups'),
+                error: function() {
+                    var errorDialog = Oskari.clazz.create('Oskari.userinterface.component.Popup');
+                    errorDialog.show(me.locale('errors.dataProvider.title'), me.locale('errors.dataProvider.message'));
+                    errorDialog.fadeout();
+                },
+                success: function(response) {
+                    me.dataProviders = [];
+                    response.organization.forEach(function(org) {
+                        me.dataProviders.push({
+                            id: org.id,
+                            name: Oskari.getLocalized(org.name)
+                        });
+                    });
+
+                    me.dataProviders.sort(function(a, b) {
+                        return Oskari.util.naturalSort(a.name, b.name);
+                    });
+                    callback();
```

Comment:
I think you should always send me.dataProviders as a parameter to the callback function and always use this function to get a reference to them. That way you can't have a situation where me.dataProviders is empty just because this function hasn't been called.